<?xml version="1.0" encoding="UTF-8"?>
<launch>
    <arg name="sim"                 default="true" />
    <arg name="dynamic_env"         default="true" />
    <arg name="collision_mapping"   default="true" />
    <arg name="2d_map"         default="2d_map" />
    <arg name="3d_map"         default="3d_map.ot" />

    <!-- start 2D dynamic obstacle fitlering -->
    <include file="$(find fetchit_mapping)/launch/filter_dynamic_obstacles.launch" if="$(arg dynamic_env)"/>

    <!-- Start 2D Mapping -->
    <include file="$(find fetchit_mapping)/launch/fetchit_mapping_2d.launch" if="$(arg dynamic_env)">
        <arg name="slam_scan_topic" value="/base_scan_static" />
    </include>
    <include file="$(find fetchit_mapping)/launch/fetchit_mapping_2d.launch" unless="$(arg dynamic_env)"/>

    <!-- Start 3D Mapping -->
    <group if="$(arg collision_mapping)">
        <include file="$(find fetchit_mapping)/launch/fetchit_octomapping.launch" if="$(arg dynamic_env)">
            <arg name="voxel_side"  value="0.025"/>
            <arg name="max_range"   value="1.82"/>
        </include>
        <include file="$(find fetchit_mapping)/launch/fetchit_octomapping.launch" unless="$(arg dynamic_env)">
            <arg name="voxel_side"  value="0.025"/>
        </include>
    </group>

    <!-- rotate slowly about z, tilting head to save the generated 2D/3D maps -->
    <group if="$(arg sim)"> <!-- params tuned to simulator -->
        <node pkg="fetchit_mapping" type="challenge_mapping_node" name="challenge_mapping" output="screen">
            <param name="look_up_angle" type="double" value="0.1" />
            <param name="look_down_angle" type="double" value="0.8" />
            <param name="tilt_period" type="double" value="1.2" />
            <param name="rotation_speed" type="double" value="0.25" />
            <param name="2d_map_name" type="str" value="$(arg 2d_map)" />
            <param name="3d_map_name" type="str" value="$(arg 3d_map)" />
        </node>
    </group>
    <group unless="$(arg sim)"> <!-- params tuned to real robot -->
        <node pkg="fetchit_mapping" type="challenge_mapping_node" name="challenge_mapping" output="screen">
            <param name="look_up_angle" type="double" value="0.1" />
            <param name="look_down_angle" type="double" value="0.8" />
            <param name="tilt_period" type="double" value="1.2" />
            <param name="rotation_speed" type="double" value="0.24" />
            <param name="2d_map_name" type="str" value="$(arg 2d_map)" />
            <param name="3d_map_name" type="str" value="$(arg 3d_map)" />
        </node>
    </group>

    <node pkg="rviz" type="rviz" name="mapping_viewer" />

</launch>